packages/core/src/lib/stream_manager.ts

```diff
@@ -0,0 +1,126 @@
+import type { PeerUpdate } from "@libp2p/interface";
+import type { Stream } from "@libp2p/interface/connection";
+import type { PeerId } from "@libp2p/interface/peer-id";
+import { Peer } from "@libp2p/interface/peer-store";
+import { Libp2p } from "@waku/interfaces";
+import { selectConnection } from "@waku/utils/libp2p";
+import debug from "debug";
+
+export class StreamManager {
+  private streamsPool: Map<string, Stream[]> = new Map();
+  private ongoingStreamCreations: Map<string, Promise<Stream>> = new Map();
+  private readonly MAX_STREAMS_PER_PEER = 3;
+  private readonly MAX_RETRIES = 3;
+
+  constructor(
+    public multicodec: string,
+    public getConnections: Libp2p["getConnections"],
+    private log: debug.Debugger
+  ) {}
+
+  private async newStream(
+    peer: Peer,
+    retries = this.MAX_RETRIES
+  ): Promise<Stream> {
+    try {
+      const connections = this.getConnections(peer.id);
+      const connection = selectConnection(connections);
+      if (!connection) {
+        throw new Error("Failed to get a connection to the peer");
+      }
+      return connection.newStream(this.multicodec);
+    } catch (error) {
+      if (retries > 0) {
+        this.log(`Retrying stream creation. Retries left: ${retries}`);
+        return this.newStream(peer, retries - 1);
+      }
+      throw error;
+    }
+  }
+
+  protected async getStream(peer: Peer): Promise<Stream> {
+    const peerIdStr = peer.id.toString();
+
+    const ongoingCreation = this.ongoingStreamCreations.get(peerIdStr);
+    if (ongoingCreation) {
+      // Wait for the ongoing stream creation to complete
+      await ongoingCreation;
+    }
+
+    const peerStreams = this.streamsPool.get(peerIdStr) || [];
+    if (peerStreams.length > 0) {
+      const stream = peerStreams.pop();
+      if (!stream) {
+        throw new Error("Failed to get a stream from the pool");
+      }
+      this.replenishStreamPool(peer);
+      return stream;
+    }
+
+    this.log(`No stream available for peer ${peerIdStr}. Opening a new one.`);
+    return this.createAndSaveStream(peer);
+  }
+
+  private async createAndSaveStream(peer: Peer): Promise<Stream> {
+    const peerIdStr = peer.id.toString();
+
+    const streamCreationPromise = (async () => {
+      const stream = await this.newStream(peer);
+      const peerStreams = this.streamsPool.get(peerIdStr) || [];
+      peerStreams.push(stream);
```

Comment:
I am pretty sure this could be your race condition. Not 100% sure how it work with the JavaScript event loop.
Maybe @weboko can advise.

In any case, I would not have a array of stream, just have 1